polygeist/test/polygeist-opt/whiletofor.mlir

```diff
@@ -31,3 +31,37 @@ module {
 // CHECK-NEXT:     }
 // CHECK-NEXT:     return
 // CHECK-NEXT:   }
+
+// -----
+
+func.func private @foo(i64, i32)
+
+// CHECK-LABEL:   func.func @extsi(
```

Comment:
Despite not applying the pattern right away, the loop is simplified.

Reply:
Can you please explain why it is not safe, and is it the same pattern that turn the while loop to scf loop, maybe after some pattern moving the `arith.extsi` outside of the before region?

Reply:
`scf.while` has two blocks:

1. The before block: this block is always evaluated at least once and has a `scf.condition` operation as a terminator, which receives an `i1` operand to determine whether the next region will execute or not and a list of additional operands which correspond to the other region's block arguments;
2. The after block: body of the loop executing if `scf.condition` is passed `true` as a first operand.

If any of those "additional arguments" passed to `scf.condition` is defined in the before region, this pattern will crash, as it is just not able to handle those. We could make this pattern more complex, but, as we already have a LICM-like pattern working on `scf.while` in this pass, I think we can just delegate the task to that pattern.

As you say, if the LICM is applied, this pattern will be applied afterwards, otherwise, it may mean this transformation cannot be performed (we will have saved a bunch of time making this pattern more complex by delegating the checking to other patterns).

Reply:
I think this approach is sound, we can definitely rely on LICM to deal with hoisting invariant values of of the while loop and just make this pattern safe.

Reply:
Do you know which pass hoisted the arguments? `LICM` operates on `LoopLikeOpInterface`, where `scf.while` doesn't implement that.

Reply:
There's [a `WhileLICM` pattern](https://github.com/intel/llvm/blob/ea569cb0a45a73d5a04cc907573b50d66759a636/polygeist/lib/Dialect/Polygeist/Transforms/CanonicalizeFor.cpp#L1534C9-L1534C9) in this pass that performs a similar transformation.